nvflare/apis/utils/reliable_message.py

```diff
@@ -0,0 +1,369 @@
+
+    def _do_request(self, request: Shareable, fl_ctx: FLContext):
+        try:
+            result = self.request_handler_f(self.topic, request, fl_ctx)
+        except Exception as e:
+            result = _error_reply(ReturnCode.EXECUTION_EXCEPTION, str(e))
+
+        # send back
+        result.set_header(HEADER_TX, self.tx_id)
+        result.set_header(HEADER_OP, OP_REPLY)
+        result.set_header(HEADER_TOPIC, self.topic)
+        self.result = result
+        self._try_reply(fl_ctx)
+
+
+class _ReplyReceiver:
+    def __init__(self, tx_id: str):
+        self.tx_id = tx_id
+        self.result = None
+        self.result_ready = threading.Event()
+
+    def process(self, reply: Shareable) -> Shareable:
+        self.result = reply
+        self.result_ready.set()
+        return make_reply(ReturnCode.OK)
+
+
+class ReliableMessage:
+
+    _topic_to_handle = {}
+    _req_receivers = {}  # tx id => receiver
+    _enabled = False
+    _executor = None
+    _query_interval = 1.0
+    _max_retries = 5
+    _max_tx_time = 300.0  # 5 minutes
+    _reply_receivers = {}  # tx id => receiver
+    _tx_lock = threading.Lock()
+    _shutdown_asked = False
+
+    @classmethod
+    def register_request_handler(cls, topic: str, handler_f):
+        if not cls._enabled:
+            raise RuntimeError("ReliableMessage is not enabled. Please call ReliableMessage.enable() to enable it")
+        if not callable(handler_f):
+            raise TypeError(f"handler_f must be callable but {type(handler_f)}")
+        cls._topic_to_handle[topic] = handler_f
+
+    @classmethod
+    def _receive_request(cls, topic: str, request: Shareable, fl_ctx: FLContext):
+        tx_id = request.get_header(HEADER_TX)
+        receiver = cls._req_receivers.get(tx_id)
+        op = request.get_header(HEADER_OP)
+        topic = request.get_header(HEADER_TOPIC)
+        if op == OP_REQUEST:
+            if not receiver:
+                handler_f = cls._topic_to_handle.get(topic)
+                if not handler_f:
+                    # no handler registered for this topic!
+                    return make_reply(ReturnCode.TOPIC_UNKNOWN)
+                receiver = _RequestReceiver(topic, handler_f, cls._executor)
+                with cls._tx_lock:
+                    cls._req_receivers[tx_id] = receiver
+            return receiver.process(request, fl_ctx)
+        elif op == OP_QUERY:
+            if not receiver:
+                return _status_reply(STATUS_NOT_RECEIVED)  # meaning the request wasn't received
+            else:
+                return receiver.process(request, fl_ctx)
+        else:
+            return make_reply(rc=ReturnCode.BAD_REQUEST_DATA)
+
+    @classmethod
+    def _receive_reply(cls, topic: str, request: Shareable, fl_ctx: FLContext):
+        tx_id = request.get_header(HEADER_TX)
+        receiver = cls._reply_receivers.get(tx_id)
+        if not receiver:
+            return make_reply(ReturnCode.OK)
```

Comment:
Why if not reply receiver is OK?

Reply:
If no receiver found, it means the request is already processed so it returns OK. This happens when the query is sent before the result is received and the 2nd reply is received after the request is already finished.